framework/src/main/java/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -1019,9 +1019,21 @@ public Void visitVariable(VariableTree node, Void p) {
 
         try {
             if (atypeFactory.getDependentTypesHelper() != null) {
-                atypeFactory
-                        .getDependentTypesHelper()
-                        .checkType(atypeFactory.getAnnotatedTypeLhs(node), node);
+                if (getCurrentPath().getParentPath() != null
+                        && getCurrentPath().getParentPath().getLeaf().getKind()
+                                == Tree.Kind.LAMBDA_EXPRESSION) {
+                    // Calling getAnnotatedTypeLhs on a lambda parameter node is possibly expensive
+                    // because caching is turned off.  This should be fixed by #979.
+                    // See https://github.com/typetools/checker-framework/issues/2853 for an
+                    // example.
+                    atypeFactory
+                            .getDependentTypesHelper()
+                            .checkType(visitorState.getAssignmentContext().second, node);
```

Comment:
I'm trying to understand the change in #2649: why isn't the assignment context always the declared type?
That is, why did #2649 change this line to `getAnnotatedTypeLhs` instead of using the assignment context?
Is the real problem maybe that the assignment context isn't set correctly to the declared type?

Reply:
> Is the real problem maybe that the assignment context isn't set correctly to the declared type?

Yes, that seems right.  I'll make that change.